hal/spi_api.h

```diff
@@ -19,220 +18,259 @@
+#endif
 
-/** Configure the SPI format
+/**
+ * Obtains the capabilities of a SPI channel.
  *
- * Set the number of bits per frame, configure clock polarity and phase, shift order and master/slave mode.
- * The default bit order is MSB.
- * @param[in,out] obj   The SPI object to configure
- * @param[in]     bits  The number of bits per frame
- * @param[in]     mode  The SPI mode (clock polarity, phase, and shift direction)
- * @param[in]     slave Zero for master mode or non-zero for slave mode
- */
-void spi_format(spi_t *obj, int bits, int mode, int slave);
-
-/** Set the SPI baud rate
+ * - returns a constant pointer to a constant structure describing the capabilities of the
+ *   related peripheral regardless of whether the device is in use or not.
  *
- * Actual frequency may differ from the desired frequency due to available dividers and bus clock
- * Configures the SPI peripheral's baud rate
- * @param[in,out] obj The SPI object to configure
- * @param[in]     hz  The baud rate in Hz
+ * @param[in]   pins    spi_pins_t structure identifying a SPI channel.
+ * @return A constant pointer to a constant structure describing the capabilities of the device.
  */
-void spi_frequency(spi_t *obj, int hz);
+const spi_capabilities_t *const spi_get_capabilities(const spi_pins_t *pins);
 
-/**@}*/
 /**
- * \defgroup SynchSPI Synchronous SPI Hardware Abstraction Layer
- * @{
- */
-
-/** Write a byte out in master mode and receive a value
+ * Initialises a SPI instance.
  *
- * @param[in] obj   The SPI peripheral to use for sending
- * @param[in] value The value to send
- * @return Returns the value received during send
- */
-int  spi_master_write(spi_t *obj, int value);
-
-/** Write a block out in master mode and receive a value
- *
- *  The total number of bytes sent and received will be the maximum of
- *  tx_length and rx_length. The bytes written will be padded with the
- *  value 0xff.
- *
- * @param[in] obj        The SPI peripheral to use for sending
- * @param[in] tx_buffer  Pointer to the byte-array of data to write to the device
- * @param[in] tx_length  Number of bytes to write, may be zero
- * @param[in] rx_buffer  Pointer to the byte-array of data to read from the device
- * @param[in] rx_length  Number of bytes to read, may be zero
- * @param[in] write_fill Default data transmitted while performing a read
- * @returns
- *      The number of bytes written and read from the device. This is
- *      maximum of tx_length and rx_length.
- */
-int spi_master_block_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, char write_fill);
-
-/** Check if a value is available to read
+ *   - initialises the pins.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if a value is available
+ * @param[in,out] obj   A spi_t instance to initialise.
+ * @param[in]     init  Initialisation parameters.
  */
-int  spi_slave_receive(spi_t *obj);
+void spi_init(spi_t *obj, const spi_init_t *init);
```

Comment:
Is init the only way to set the frequency and other settings? This should support at least changing the frequency after init, and probably the other settings as well. I know the current SD card driver connects at 100KHz and then once reading the supported speed switches to a higher frequency.

Reply:
Indeed, it was agreed with @bulislaw that as there is no use-case where frequency & settings are changed on-the-fly/mid transfer (such operation only happen while `SS` is **not** asserted) and that it only happens at the early stage of the channel lifetime it is fine to free the channel & re-initialise it. 
It is unlikely to be a problem for the slave as pull-ups should be provided externally for the sake of startup/reset safety.